libnavigation-core/src/main/java/com/mapbox/navigation/core/history/HistoryRecorderController.kt

```diff
@@ -0,0 +1,69 @@
+package com.mapbox.navigation.core.history
+
+import com.mapbox.base.common.logger.Logger
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
+import com.mapbox.navigation.base.options.NavigationOptions
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigator.HistoryRecorderHandle
+
+/**
+ * Internal class that surfaces error messages and holds onto
+ * nav-native's [HistoryRecorderHandle].
+ */
+internal class HistoryRecorderController(
+    navigationOptions: NavigationOptions,
+    val logger: Logger
+) {
+    private val historyRecorderOptions = navigationOptions.historyRecorderOptions
+    private val historyFiles = HistoryFiles(navigationOptions.applicationContext, logger)
+
+    var historyRecorderHandle: HistoryRecorderHandle? = null
+        internal set
+
+    /**
+     * Used to expose a way for external developers to
+     * save history files via [MapboxNavigation.historyRecorder].
+     *
+     * This is needed to support [MapboxNavigation.resetTripSession]
+     */
+    val historyRecorder = object : MapboxHistoryRecorder {
+
+        /**
+         * @see MapboxHistoryRecorder.fileDirectory
+         */
+        override fun fileDirectory(): String? {
+            return if (historyRecorderOptions.enabled) {
+                historyFiles.absolutePath(historyRecorderOptions)
+            } else {
+                null
+            }
+        }
+
+        /**
+         * @see MapboxHistoryRecorder.saveHistory
+         */
+        override fun saveHistory(result: SaveHistoryCallback) {
+            historyRecorderHandle?.dumpHistory { filePath: String? ->
+                result.onSaved(filePath)
+            }
+            check(historyRecorderOptions.enabled) {
+                logger.e(
+                    Tag("MbxHistoryRecorderOptions"),
+                    Message("You must enable HistoryRecorderOptions in order to save history")
+                )
+            }
+            historyRecorderHandle?.apply {
+                dumpHistory { filePath: String? ->
+                    result.onSaved(filePath)
+                }
+            } ?: run {
+                logger.w(
+                    Tag("MbxHistoryRecorderOptions"),
+                    Message("dumpHistory failed - The history recorder is not initialized")
+                )
+                result.onSaved(null)
+            }
```

Comment:
Not sure if I understand this, why are we trying to call `dumpHistory` twice? 🤔

Reply:
testing locally and pushed when i went for family dinner 💩